### PR TITLE
Make the various usage of ConfigureServices more consistent

### PR DESCRIPTION
**Pull request by mattleibow**
### Description of Change ###

Needs to be merged before #606 

As I was adding "feature" sets to the builder (like the fonts, lifecycle events, essentials), I noticed a few things:
1. The root level (builder.*) was getting a bit bloated with a few overloads for each "feature"
2. There were a few verbs being used (Add, Register, Use, Configure)
3. I kept ading members to the `IAppHostBuilder` interface

## Bloating of Root Builder

To fix 1. I tried to find a way to group each selection of extensions for a "feature". One of these ways was to follow the pattern that already existed with the `ConfigureServices` and `ConfigureAppConfiguration`. Instead of having all the `AddTransient` and `AddSingleton` extensions directly on the `IAppHostBuilder`, it is under the `ConfigureServices`.

```csharp
builder
    .ConfigureServices((hostBuilderContext, serviceCollection) =>
    {
        serviceCollection.AddTransient<T>();
        serviceCollection.AddSingleton<T>();
    })
    .ConfigureAppConfiguration((hostBuilderContext, configurationBuilder) =>
    {
        configurationBuilder.AddInMemoryCollection();
    })
```

It seemed natural to follow this with the handlers and fonts:

```csharp
builder
    .ConfigureMauiHandlers((hostBuilderContext, handlersCollection) =>
    {
        handlersCollection.AddHandler<IView, ViewHandler>();
    })
    .ConfigureFonts((hostBuilderContext, fontCollection) =>
    {
        fontCollection.AddFont("Dokdo-Regular.ttf", "Dokdo");
    })
```

For things like the Compatibility renderer shims, instead of adding another method to the root builder, we can also add a method that follows with what we have here already:

```csharp
builder
    .ConfigureMauiHandlers((hostBuilderContext, handlersCollection) =>
    {
        handlersCollection.AddCompatibilityRenderer<View, ViewRenderer>();
        handlersCollection.AddHandler<IView, ViewHandler>();
    })
```

After doing all this, I still wanted to reduce the amount of typing to do simple things, like adding a handler or enabling all the renderers. For these cases, it is very easy to create a wrapper for the specific `Configure*` method and add a root-level extension:

```csharp
builder
    .UseDefaultCompatibilityRenderers()
    .UseMauiHandler<IView, ViewHandler>()
````

_I left `UseMauiHandler` out for now as this PR didn't actually need it and we can add it later._

## Varying Verb Usage

The other issue was the different verbs that were being used and the meaning might get confusing if we introduce new ones as well as chanhe what they do:

```csharp
builder
    // Register*
    .RegisterCompatibilityRenderers()
    .RegisterCompatibilityRenderer<View, ViewRenderer>()
    .RegisterHandler<IView, ViewHandler>()
    // Configure*
    .ConfigureServices()
    .ConfigureHandlers()
    // Use*
    .UseFonts()
    .UseMauiHandlers()
    // Add*
    .AddEssentials()
```

To "solve" this (as I still am not happy) I wend the way of what was already there: using the `Configure*` verb to represent a way to dynamically "configure" a feature and the `Use*` verb to represent a way to turn a feature on and off.

For the "configurable" things we have the `Configure*` prefix:

```csharp
builder
    .ConfigureServices((_, services) =>
    {
        services.AddSingleton<T>();
    })
    .ConfigureHandlers((_, handlers) =>
    {
        handlers.AddHandler<T>();
        handlers.AddCompatibilityRenderer<T, T>();
    })
    .ConfigureFonts((_, fonts) =>
    {
        fonts.AddFont("");
        fonts.AddEmeddedResourceFont("");
    })
```

As you can also see, once inside the scope of a configure, we can use the `Add*` prefix to add things to a sort of collection/builder that is provided.

For the "just use" things we have the `Use*` prefix:

```csharp
builder
    .UseMauiApp<TApp>()
    .UseDefaultCompatibilityRenderers()
    .UseServiceProviderFactory()
```

> However, this is where things start to get messy. All things look nice so far, but what about things that are configurable, features and are not collections?

An example of something that does not fit is essentials because it is both a feature and a configurable. I could do this:

```csharp
builder
    .UseEssentials()
```

But what if we want to set some options on this:

```csharp
builder
    .ConfigureEssentials((_, essentials) =>
    {
        essentials
            .UseVersionTracking()
            .UseMapServiceToken("")
            .AddAppAction("", "")
            .OnAppAction(action => { })
    })
```

## Adding to `IAppHostBuilder`

The other thing that we can't sustain is the ability to add features to the `IAppHostBuilder` interface. Each time we add something, it breaks ABI. And, at the end of the day, we are going to have so many things there and no way to remove anything as that is now also breaking.

The reason we need to add features here is for the fact that you really want to first collect all the parts and then finally in `Build()` create the service. An example is with the handlers. Each time we call `ConfigureHandlers`, we are really wanting to add the delegate to a collection. Then, at the `Build()` step, we collect all the handlers and register them.

I also noticed with things like lifecycle events, this is also a feature that is needed. Each plugin will register events and then we would expect them all to get consolidated into a single `ILifecycleEventService` or something. There is even the case for fonts where we want to register a collection of fonts, just like the services/handlers, but also at build time be able to access the service provider and register things.

To solve this, I added a new interface `IMauiServiceBuilder` which really is a mini `Startup` class:

```csharp
public interface IMauiServiceBuilder
{
    // Logic that runs before the service provider is created
    // and can be used to configure and add services to the
    // service collection.
    void ConfigureServices(IServiceCollection services);

    // Logic that runs after the service provider has been
    // created and now can be used to pull things from it
    // and actually set up the system.
    void Configure(IServiceProvider services);
}
```

The reason that I say this is a mini `Startup` class is that it really does fit into the pattern that was set up for ASP.NET. In reality, out `Startup` class is more the `Program.cs` file that allows for the configuration of the app host. Then, the configuration of the middleware in ASP.NET is actually different to what we need. We don't need to set up a render pipeline but rather have a way for plugin authors to hook into our `Build()` system.

One benefit of this is that each time the user requests a builder object for configuration, they get the _same_ object allowing for not only compositional configuration but also the plugin author can hook up things without having to worry about multi-registration.

To demonstrate how this new interface works, we can hane a look at fonts:

```csharp
class FontCollectionBuilder : Collection<FontDescriptor>, IMauiServiceBuilder
{
    public void Add(string filename, string alias)
    {
        this.Add(new FontDescriptor(filename, alias));
    }

    public void ConfigureServices(IServiceCollection services)
    {
        services.AddSingleton<IEmbeddedFontLoader, EmbeddedFontLoader>();
        services.AddSingleton<IFontRegistrar, FontRegistrar>();
        services.AddSingleton<IFontManager, FontRegistrar>();
    }

    public void Configure(IServiceProvider services)
    {
        var fontRegistrar = services.GetService<IFontRegistrar>();
        foreach (var font in this)
        {
            fontRegistrar.Register(font.Filename, font.Alias);
        }
    }
}
```

If the user was to use this to set up the app:

```csharp
builder
    .ConfigureFonts((_, fonts) => fonts.Add("Normal-Regular.ttf", "Normal"))
    .ConfigureFonts((_, fonts) => fonts.Add("Cute-Regular.ttf", "Cute"))
    .ConfigureServices((_, services)=> services.AddSingleton<IFontRegistrar, CustomFontRegistrar>())
```

The effect of all this would be this:

1. The "Normal" font registration delegate is added to an internal list.
2. The "Cute" font registration delegate is added to an internal list.
3. The `CustomFontRegistrar` registration is added to an internal list.
4. `Build()` is called by the system.
5. A new `FontCollectionBuilder` object is constructed.
6. Both font registration delegates are invoked passing in the `FontCollectionBuilder` instance.
7. The font builder's `ConfigureServices` method is called and the plugin registers its own services.
8. The app host bulder's service collection is configured with all the `ConfigureServices` calls resulting in the user's registration of `CustomFontRegistrar` overwriting the plugin author's implementation of `IFontRegistrar`.
9. The service provider is constructed.
10. The font builder's `Configure` method is called and the plugin requests the `IFontRegistrar` registrar which will return a new instance of the user's `CustomFontRegistrar`. This object is then used and all the fonts are registered with it.

### Additions made ###

Quite a few. And lots of changes.

### PR Checklist ###

- [x] Targets the correct branch 
- [x] Tests are passing (or failures are unrelated)
- [x] Implements tests

### CODE REVIEW

File: src/Core/src/Hosting/Internal/AppHost.cs [hunk 1]
```diff
@@ -6,21 +6,21 @@
 
 namespace Microsoft.Maui.Hosting.Internal
 {
-	internal class AppHost : IAppHost, IAsyncDisposable
+	class AppHost : IAppHost, IAsyncDisposable
```
Comment (rmarinho) [inline src/Core/src/Hosting/Internal/AppHost.cs:9]: if it's not internal maybe shouldn't be on internal namespace ?

File: src/Core/src/Hosting/Internal/AppHost.cs [hunk 2]
```diff
@@ -6,21 +6,21 @@
...
 	{
 		readonly ILogger<AppHost>? _logger;
 
 		public AppHost(IServiceProvider services, ILogger<AppHost>? logger)
 		{
 			Services = services ?? throw new ArgumentNullException(nameof(services));
-			Handlers = Services.GetRequiredService<IMauiHandlersServiceProvider>();
+			Handlers = Services.GetService<IMauiHandlersServiceProvider>();
```
Comment (rmarinho) [inline src/Core/src/Hosting/Internal/AppHost.cs:16]: why are we changing this? i think we really should crash if there's no `IMauiHandlersServiceProvider`

File: src/Core/src/Hosting/Internal/AppHostEnvironment.cs [hunk 3]
```diff
@@ -3,7 +3,7 @@
 
 namespace Microsoft.Maui.Hosting.Internal
 {
-	internal class AppHostEnvironment : IHostEnvironment
+	class AppHostEnvironment : IHostEnvironment
```
Comment (rmarinho) [inline src/Core/src/Hosting/Internal/AppHostEnvironment.cs:6]: Why this change? if not internal maybe move it out of this namespace/folder

File: src/Core/src/Hosting/Internal/MauiServiceProviderFactory.cs [hunk 4]
```diff
@@ -3,7 +3,7 @@
 
 namespace Microsoft.Maui.Hosting.Internal
 {
-	internal class MauiServiceProviderFactory : IServiceProviderFactory<IServiceCollection>
+	class MauiServiceProviderFactory : IServiceProviderFactory<IServiceCollection>
```
Comment (rmarinho) [inline src/Core/src/Hosting/Internal/MauiServiceProviderFactory.cs:6]: Same comment as above, if it's on the namespace and folder it should be internal, or else move it out.

File: src/Controls/samples/Controls.Sample/Startup.cs [hunk 5]
```diff
@@ -45,7 +45,7 @@ public void Configure(IAppHostBuilder appBuilder)
 				})
 				.UseMauiServiceProviderFactory(true)
 				//.UseServiceProviderFactory(new DIExtensionsServiceProviderFactory())
-				.ConfigureServices((hostingContext, services) =>
+				.ConfigureServices(services =>
```
[the remark this answers is not in the document]
Comment (mattleibow) [reply src/Controls/samples/Controls.Sample/Startup.cs:48]: I agree. This is just the simpler overload that I am using. The actual interface member has it still - since it is in the base. I just added an extension method to make things less intense for the new folks.

File: src/Core/src/Fonts/Hosting/AppHostBuilderExtensions.cs [hunk 6]
```diff
@@ -0,0 +1,53 @@
+using System;
+using Microsoft.Extensions.DependencyInjection;
+using Microsoft.Extensions.Hosting;
+using Microsoft.Maui.Hosting.Internal;
+
+namespace Microsoft.Maui.Hosting
```
[the remark this answers is not in the document]
Comment (mattleibow) [reply src/Core/src/Fonts/Hosting/AppHostBuilderExtensions.cs:6]: Yep. Moving.

File: src/Compatibility/Core/src/AppHostBuilderExtensions.cs [hunk 7]
```diff
@@ -26,80 +49,23 @@ public static IAppHostBuilder RegisterCompatibilityForms(this IAppHostBuilder bu
 			return builder;
 		}
 
-		public static IAppHostBuilder RegisterCompatibilityRenderers(this IAppHostBuilder builder)
+		public static IAppHostBuilder UseDefaultCompatibilityRenderers(this IAppHostBuilder builder)
```
[the remark this answers is not in the document]
Comment (mattleibow) [reply src/Compatibility/Core/src/AppHostBuilderExtensions.cs:52]: Nope. Xapping.

File: src/Controls/samples/Controls.Sample.SingleProject/Startup.cs [hunk 8]
```diff
@@ -9,7 +9,8 @@ public class Startup : IStartup
 		public void Configure(IAppHostBuilder appBuilder)
 		{
 			appBuilder
-				.RegisterCompatibilityRenderers()
+				.UseDefaultCompatibilityRenderers()
+				.UseCompatibilityForms()
```
[the remark this answers is not in the document]
Comment (mattleibow) [reply src/Controls/samples/Controls.Sample.SingleProject/Startup.cs:13]: Yep. No need for the dupes.

File: src/Core/src/Hosting/AppHostBuilder.cs [hunk 9]
```diff
@@ -29,25 +28,23 @@ public class AppHostBuilder
...
 
 		public static IAppHostBuilder CreateDefaultAppBuilder()
 		{
 			var builder = new AppHostBuilder();
 
-			builder.UseMauiHandlers();
-			builder.UseFonts();
+			builder.UseMauiServiceProviderFactory(false);
+
+			builder.UseDefaultMauiHandlers();
```
[the remark this answers is not in the document]
Comment (mattleibow) [reply src/Core/src/Hosting/AppHostBuilder.cs:39]: Who knows. I sometimes do crazy things. Cleaning up.

File: src/Core/src/Hosting/AppHostBuilder.cs [hunk 10]
```diff
@@ -184,18 +185,15 @@ void CreateServiceProvider(IServiceCollection services)
 
 		void BuildAppConfiguration()
 		{
-			if (_hostBuilderContext == null)
-				return;
```
[the remark this answers is not in the document]
Comment (mattleibow) [reply src/Core/src/Hosting/AppHostBuilder.cs:188]: Yup. Fixing.

File: src/Core/src/Hosting/AppHostBuilder.cs [hunk 11]
```diff
@@ -204,45 +202,40 @@ IServiceProvider ConfigureContainerAndGetProvider(IServiceCollection services)
 
 			foreach (IConfigureContainerAdapter containerAction in _configureContainerActions)
 			{
-				if (_hostBuilderContext != null)
```
[the remark this answers is not in the document]
Comment (mattleibow) [reply src/Core/src/Hosting/AppHostBuilder.cs:207]: Yup. Fixing.

File: src/Core/src/Hosting/Internal/AppHost.cs (hunk 1 above)
Comment (mattleibow) [reply src/Core/src/Hosting/Internal/AppHost.cs:9]: Just removing the internal as it is the default.

File: src/Core/src/Hosting/Internal/AppHostEnvironment.cs (hunk 3 above)
Comment (mattleibow) [reply src/Core/src/Hosting/Internal/AppHostEnvironment.cs:6]: This is just removing the keyword as internal is default.

File: src/Core/src/Hosting/Internal/MauiServiceProviderFactory.cs (hunk 4 above)
Comment (mattleibow) [reply src/Core/src/Hosting/Internal/MauiServiceProviderFactory.cs:6]: This is just removing the keyword as internal is default.

File: src/Core/src/Hosting/Internal/AppHost.cs (hunk 2 above)
Comment (mattleibow) [reply src/Core/src/Hosting/Internal/AppHost.cs]: Yup.

**Comment by mattleibow** [comment]
Thanks for the review. I made all the changes.

File: eng/Microsoft.Extensions.targets [hunk 12]
```diff
@@ -31,6 +35,10 @@
         Update="Microsoft.Extensions.Logging"
         Version="$(_MicrosoftHostingVersion)"
     />
+    <PackageReference
```
[the remark this answers is not in the document]
Comment (mattleibow) [reply eng/Microsoft.Extensions.targets:38]: This is just an update, so it will set the value, only if it exists.

[the remark this answers is not in the document]
Comment (mattleibow) [reply eng/Microsoft.Extensions.targets:38]: I thought it was. Let me cleanup stuff.

File: eng/Microsoft.Extensions.targets [hunk 13]
```diff
@@ -19,6 +19,10 @@
         Update="Microsoft.Extensions.DependencyInjection"
         Version="$(_MicrosoftDependencyInjectionVersion)"
     />
+    <PackageReference
```
[the remark this answers is not in the document]
Comment (rmarinho) [reply eng/Microsoft.Extensions.targets:22]: this is for the sample . in the sample we test against the Microsoft.Extensions.DependencyInjection

File: .nuspec/Microsoft.Maui.Controls.MultiTargeting.targets [hunk 14]
```diff
@@ -32,10 +32,10 @@
   <ItemGroup Condition="$(TargetFramework.Contains('-windows')) != true ">
     <Compile Remove="**\*.Windows.cs" />
     <None Include="**\*.Windows.cs" />
-    <Compile Remove="**\*.xaml.cs" />
-    <None Include="**\*.xaml.cs" />
-    <Compile Remove="**\*.xaml" />
-    <None Include="**\*.xaml" />
+    <Compile Remove="**\*.Windows.xaml.cs" />
```
[the remark this answers is not in the document]
Comment (mattleibow) [reply .nuspec/Microsoft.Maui.Controls.MultiTargeting.targets]: Right now, this file is only used by Core and DeviceTests which do not have any xaml. I removed it for now so it doesn't cause issue later on if I did it wrong.

**Review by Redth — APPROVED** [review]
